backend/de.metas.material/cockpit/src/main/java/de/metas/material/cockpit/availableforsales/AvailableForSalesService.java

```diff
@@ -0,0 +1,249 @@
+	{
+		final AvailableForSalesQuery availableForSalesQuery = enqueueAvailableForSalesRequest.getAvailableForSalesQuery();
+
+		Loggables.withLogger(logger, Level.DEBUG).addLog("ProductId: {} and AttributesKey: {} enqueued to be synced.",
+														 availableForSalesQuery.getProductId(),
+														 availableForSalesQuery.getStorageAttributesKeyPattern().getSqlLikeString());
+
+		syncProductDebouncer.add(enqueueAvailableForSalesRequest);
+	}
+
+	public void syncAvailableForSalesForProduct(@NonNull final Product product)
+	{
+		final ImmutableList<AvailableForSalesQuery> availableForSalesQueries = buildAvailableForSalesQueries(product);
+
+		for (final AvailableForSalesQuery availableForSalesQuery : availableForSalesQueries)
+		{
+			syncAvailableForSalesTable(availableForSalesQuery);
+		}
+	}
+
+	@NonNull
+	public AvailableForSalesMultiResult computeAvailableForSales(@NonNull final AvailableForSalesMultiQuery query)
+	{
+		return availableForSalesRepository.computeAvailableForSales(query, productBL::getStockUOMId);
+	}
+
+	@NonNull
+	private ImmutableList<AvailableForSalesQuery> buildAvailableForSalesQueries(@NonNull final Product product)
+	{
+		final OrgId orgId = product.getOrgId();
+
+		if (OrgId.ANY.equals(orgId))
+		{
+			return orgDAO.retrieveClientOrgs(Env.getAD_Client_ID())
+					.stream()
+					.map(I_AD_Org::getAD_Org_ID)
+					.map(OrgId::ofRepoId)
+					.map(currentOrgId -> createAvailableForSalesQuery(product, currentOrgId))
+					.filter(Optional::isPresent)
+					.map(Optional::get)
+					.collect(ImmutableList.toImmutableList());
+		}
+		else
+		{
+			return createAvailableForSalesQuery(product, orgId)
+					.map(ImmutableList::of)
+					.orElseGet(ImmutableList::of);
+		}
+	}
+
+	private void syncAvailableForSales(@NonNull final Collection<EnqueueAvailableForSalesRequest> requests)
+	{
+		if (requests.isEmpty())
+		{
+			Loggables.withLogger(logger, Level.DEBUG).addLog("SyncAvailableForSalesRequest list is empty! No action is performed!");
+			return;
+		}
+
+		for (final EnqueueAvailableForSalesRequest request : requests)
+		{
+
+			try (final IAutoCloseable autoCloseable = Env.switchContext(request.getCtx()))
+			{
+				syncAvailableForSalesTable(request.getAvailableForSalesQuery());
+			}
+		}
+	}
+
+	private void syncAvailableForSalesTable(@NonNull final AvailableForSalesQuery availableForSalesQuery)
+	{
+		final ImmutableList<AvailableForSalesResult> availableForSalesResults = computeAvailableForSales(AvailableForSalesMultiQuery.of(availableForSalesQuery))
+				.getAvailableForSalesResults();
+
+		final ImmutableList<I_MD_Available_For_Sales> availableForSalesRecords = availableForSalesRepository.getRecordsByQuery(availableForSalesQuery);
+
+		final ImmutableMap<I_MD_Available_For_Sales, AvailableForSalesResult> recordsToBeUpdated =
+				mapRecordsWithResultBasedOnASI(availableForSalesRecords, availableForSalesResults);
+
+		recordsToBeUpdated
+				.forEach((record, result) -> {
```

Comment:
i don't really see the point of using forEach here, but since i can set my breakpoint in this lambda, it's fine with me :-)